server/routes.go

```diff
@@ -168,93 +168,308 @@ func GenerateHandler(c *gin.Context) {
+		}
+	}()
+
+	if req.Stream != nil && !*req.Stream {
+		// Wait for the channel to close
+		var r api.GenerateResponse
+		for resp := range ch {
+			var ok bool
+			if r, ok = resp.(api.GenerateResponse); !ok {
+				c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+				return
+			}
+		}
+		r.Response = generated.String()
+		c.JSON(http.StatusOK, r)
+		return
+	}
+
+	streamResponse(c, ch)
+}
+
+func ChatHandler(c *gin.Context) {
+	loaded.mu.Lock()
+	defer loaded.mu.Unlock()
+
+	checkpointStart := time.Now()
+
+	var req api.ChatRequest
+	err := c.ShouldBindJSON(&req)
+	switch {
+	case errors.Is(err, io.EOF):
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": "missing request body"})
+		return
+	case err != nil:
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	// validate the request
+	switch {
+	case req.Model == "":
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": "model is required"})
+		return
+	case len(req.Format) > 0 && req.Format != "json":
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": "format must be json"})
+		return
+	}
+
+	sessionDuration := defaultSessionDuration
+	model, err := GetModel(req.Model)
+	if err != nil {
+		var pErr *fs.PathError
+		if errors.As(err, &pErr) {
+			c.JSON(http.StatusNotFound, gin.H{"error": fmt.Sprintf("model '%s' not found, try pulling it first", req.Model)})
+		} else {
+			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		}
+		return
+	}
+
+	workDir := c.GetString("workDir")
+	if err := load(c.Request.Context(), workDir, model, req.Options, sessionDuration); err != nil {
+		if errors.Is(err, api.ErrInvalidOpts) {
+			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		} else {
+			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+		}
+		return
+	}
+
+	// an empty request loads the model
+	if len(req.Messages) == 0 {
+		c.JSON(http.StatusOK, api.ChatResponse{CreatedAt: time.Now().UTC(), Model: req.Model, Done: true})
+		return
+	}
+
+	checkpointLoaded := time.Now()
+
+	// build the prompt from the list of messages
+	var promptVars []PromptVars
+	currentVars := PromptVars{}
+
+	for _, m := range req.Messages {
+		switch m.Role {
+		case "system":
+			if currentVars.Prompt != "" || currentVars.System != "" {
+				promptVars = append(promptVars, currentVars)
+				currentVars = PromptVars{}
+			}
+			currentVars.System = m.Content
+		case "user":
+			if currentVars.Prompt != "" || currentVars.System != "" {
+				promptVars = append(promptVars, currentVars)
+				currentVars = PromptVars{}
+			}
+			currentVars.Prompt = m.Content
+		case "assistant":
+			currentVars.Response = m.Content
+			promptVars = append(promptVars, currentVars)
+			currentVars = PromptVars{}
+		default:
+			c.JSON(http.StatusBadRequest, gin.H{"error": fmt.Sprintf("invalid role: %s, role must be one of [system, user, assistant]", m.Role)})
 			return
 		}
+	}
```

Comment:
Could this go into a `model.ChatPrompt(req.Messages)` (that could, in itself call `model.Prompt()` in a loop to build the final template?).